### DISCUSSION TITLE
EachBean injecting from other EachBean

### DISCUSSION THREAD

**Discussion by cancech** [micronaut-projects/micronaut-core#6630] [Q&A] [answered]
By way of context, imagine that there are two classes `Vehicle` and `Radio`. Each `Vehicle` contains a single `Radio` and we want this `Radio` to be unique for each `Vehicle` (so can't just treat it as a Singleton) and each instance once created is to be unique for the `Vehicle` and be accessible/injectable for the duration of the execution of the application (so can't just treat it as a Prototype as that would create a new copy each time I inject it). After reading through the documentation it would seem that the `EachBean` allows me to create unique copies of each in the manner I desire, so I can now define the `Radio` class

```Java
@EachBean(MyDuplicationConfig.class)
public class Radio {
    Radio(MyDuplicationConfig conf) {
        // Do something
    }
}
```

But now the question is, how do I get this `Radio` bean into the `Vehicle`? After experimenting, the best I can think of would be using a Factory to pull in the created `Radio` instances and passing it into the `Vehicle` constructor in that manner

```Java
@Factory
public class VehicleFactory {
    @Inject
    @Named("copy1")
    Radio radio1;

    @Inject
    @Named("copy2")
    Radio radio2;

    @EachBean(MyDuplicationConfig.class)
    Vehicle createVehicle(MyDuplicationConfig conf) {
        if (conf.getName().equals("copy1"))
            return new Vehicle(radio1);

        return new Vehicle(radio2);
    }
}
```

And that seems to work, however then you end up with some duplication (need to inject the radio twice) and you also now need to be extra careful when adding/removing/renaming the names of the duplicate instances. From what I've seen it would appear that there is a mechanism for getting around this limitation by using the ApplicationContext directly.

```Java
@Factory
public class VehicleFactory {
    @Inject
    ApplicationContext ctx;

    @EachBean(MyDuplicationConfig.class)
    Vehicle createVehicle(MyDuplicationConfig conf) {
       return new Vehicle(ctx.getBean(Radio.class, Qualifiers.byName(prop.getName())));
    }
}
```

Much more concise, but is this safe? It's working in the limited testing I've performed, but my fear is that the `createVehicle` might be triggered prior to the `Radio` instances being created.

I believe that the simplest and safest approach would be to allow `Vehicle` to just inject the `Radio` instance directly.

```Java
@EachBean(MyDuplicationConfig.class)
public class Vehicle {
    Vehicle(MyDuplicationConfig conf, @Named(conf.getName()) Radio radio) {
        // Do something
    }
}
```

Can this be done? Is it possible for class annotated as EachBean to inject from another class annotated as EachBean? The above doesn't compile and variations on that idea that I've tried either don't compile or don't work.

What would be the suggest solution of achieving this result?

**Comment by graemerocher** [comment, marked as answer]
We use this pattern all the time in Micronaut modules, try use `@Parameter Radio radio`

**Reply by cancech** [reply]
This is exactly what I was looking for, thanks!
